test/base/TypeDemangler.test.cc

```diff
@@ -0,0 +1,90 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file TypeDemangler.test.cc
+//---------------------------------------------------------------------------//
+#include "base/TypeDemangler.hh"
+
+#include "celeritas_test.hh"
+
+using celeritas::demangled_typeid_name;
+using celeritas::TypeDemangler;
+
+//---------------------------------------------------------------------------//
+
+namespace tdtest
+{
+template<class T>
+struct FlorbyDorb
+{
+};
+
+struct Zanzibar
+{
+};
+
+struct JapaneseIsland
+{
+    virtual ~JapaneseIsland() {}
+};
+
+struct Honshu : public JapaneseIsland
+{
+};
+
+struct Hokkaido : public JapaneseIsland
+{
+};
+} // namespace tdtest
+
+//---------------------------------------------------------------------------//
+// TESTS
+//---------------------------------------------------------------------------//
+
+TEST(TypeDemanglerTest, demangled_typeid_name)
+{
+    std::string int_type = demangled_typeid_name(typeid(int).name());
+    std::string flt_type = demangled_typeid_name(typeid(float).name());
+
+    EXPECT_NE(int_type, flt_type);
+
+#ifdef __GNUG__
+    EXPECT_EQ("int", int_type);
+    EXPECT_EQ("float", flt_type);
+#endif
+}
+
+TEST(TypeDemanglerTest, static_types)
+{
+    using namespace tdtest;
+
+    TypeDemangler<FlorbyDorb<Zanzibar>> demangle_type;
+
+    std::string fdz_type = demangle_type();
+    EXPECT_NE(fdz_type, TypeDemangler<FlorbyDorb<Hokkaido>>()());
+
+#ifdef __GNUG__
+    EXPECT_EQ("tdtest::FlorbyDorb<tdtest::Zanzibar>", fdz_type);
```

Comment:
Why do we expect the demangled name to be different (for this "simple" case) on different platforms?

Reply:
Just in case anyone wants to compile with intel or windows or anything else ;). The underlying `demangle_typeid` implementation relies on `cxxabi.h` which isn't standard.

Reply:
You can find a windows implementation at: https://github.com/root-project/root/blob/889defb6f0f487293da38a5921ab316a7eda2c7f/core/foundation/inc/TClassEdit.h#L224

> Just in case anyone wants to compile with ...

fair enough, I missed the fact that the fall-back behavior is to not demangle :)

Reply:
Ah, thanks for the link! I don't expect we'll ever be building on windows (or certainly debugging on windows) but that's a good reference to have.